### PR TITLE
2-qubit matrix decomposition into ion trap gates

### PR DESCRIPTION
**Pull request by mlihc15**
Adding ion_decomposition.py and ion_decomposition_test.py. 
cirq.two_qubit_matrix_to_ion_operations converts two-qubit matrix to ion trap gates. This method is essentially using KAK decomposition and translates all non local interactions into cirq.MS gate, which is the native 2-qubit gate in ion traps.

### CODE REVIEW

File: cirq/ion/ion_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                                     kak: linalg.KakDecomposition,
+                                     tolerance: float = 1e-8
+                                     ) -> List[ops.Operation]:
+    """Assumes that the decomposition is canonical."""
+    b0, b1 = kak.single_qubit_operations_before
+    pre = [_do_single_on(b0, q0, tolerance), _do_single_on(b1, q1, tolerance)]
+    a0, a1 = kak.single_qubit_operations_after
+    post = [_do_single_on(a0, q0, tolerance), _do_single_on(a1, q1, tolerance)]
+
+    return list(ops.flatten_op_tree([
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition.py]: Check out `cirq/optimizers/two_qubit_decompositions.py::_cleanup_operations`, which is now called at the end of the `_kak_decomposition_to_operations` method to clear out unnecessary extra single qubit gates. You should define and use a similar method here.

Comment (mlihc15) [reply cirq/ion/ion_decomposition.py]: I was going to ask you about how to clean up single qubit gates. Thanks for the tip!

**Comment by Strilanc** [comment]
Still waiting for some example circuit printouts (not in the code, but here on github).

**Comment by mlihc15** [comment]
> Still waiting for some example circuit printouts (not in the code, but here on github).

Sorry I forgot about the first comment. The circuit optimization is not ideal.
Example: cirq.unitary(cirq.MS(np.pi/4)) = 
[cirq.X.on(cirq.NamedQubit('q0')), cirq.X.on(cirq.NamedQubit('q1')), cirq.MS(-0.5*np.pi/2).on(cirq.NamedQubit('q0'), cirq.NamedQubit('q1'))]

This is because the KAK decomposition needs to first transform MS(np.pi/4) into XX interaction and single qubit gates.

Any idea on how to absorb the single qubit Xs into MS?

Also the cleanup routine doesn't work too well. We need to make some new optimization functions that understands MS gates (instead of CZs).

The decomposition works fine, its just the optimization is very tricky.

**Comment by Strilanc** [comment]
Hmmmmm, I see.

```
In [8]: cirq.Circuit.from_ops(cirq.ion.two_qubit_matrix_to_ion_operations(a, b, cirq.unitary(cirq.MS(np.pi*0.25))))                                                                                                
Out[8]: 
0: ───Z───MS(-0.25π)───Y───
          │
1: ───Z───MS(-0.25π)───Y───
```

We can add a pass akin to the phased Pauli ejection optimization to fix this. Basically we have the following rule: a Z before the MS gate can be passed over the MS(pi/4) or MS(-pi/4) gate by turning the Z into a Y and adding an X on the other wire. An equivalent rule exists for Y: passing over transforms the Y into a Z and kicks out an X on the other line.

This works because the MS gate at those angles has the following form:

image

Passing the Z over the sqrt(X) is what turns the Z into a Y, and passing it over the XNOT is what kicks out the other X. Technically you should transform into -Y instead of Y in some cases, but that's only if you care about global phase.

I would copy paste the `EjectPhasedPaulis` class and cut down the rules to just that Y and Z thing.

File: cirq/ion/ion_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            construction.
+
+    Returns:
+        A list of operations implementing the matrix.
+    """
+    kak = linalg.kak_decomposition(mat,
+                                   linalg.Tolerance(atol=tolerance))
+    operations = _kak_decomposition_to_operations(q0,
+        q1, kak, tolerance)
+    # print('\n', "not optimized")
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition.py]: Leftover debug lines

File: cirq/ion/ion_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    # print(operations)
+    # print(len(operations))
+    return _cleanup_operations(operations)
+
+
+def _cleanup_operations(operations: List[ops.Operation]):
+    circuit = circuits.Circuit.from_ops(operations)
+    optimizers.merge_single_qubit_gates.\
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition.py]: Is `merge_single_qubit_gates_into_phased_x_z` not exposed by `optimizers`? Most of the things you're digging down for should be at the top level.

File: cirq/ion/ion_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+
+def _non_local_part(q0: ops.QubitId,
+                    q1: ops.QubitId,
+                    interaction_coefficients: Tuple[float, float, float],
+                    tolerance: float = 1e-8):
+    """Yields non-local operation of KAK decomposition."""
+
+    x, y, z = interaction_coefficients
+    # print(interaction_coefficients)
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition.py]: Debug line

File: cirq/ion/ion_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+    return[
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition.py]: Nit: missing space

File: cirq/ion/ion_decomposition_test.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+])
+def test_two_to_ops(
+        max_ms_depth,
+        effect):
+    q0 = cirq.NamedQubit('q0')
+    q1 = cirq.NamedQubit('q1')
+
+    operations = cirq.two_qubit_matrix_to_ion_operations(
+        q0, q1, effect)
+    # print("optimized")
```
Comment (Strilanc) [inline cirq/ion/ion_decomposition_test.py]: Debug lines

**Comment by mlihc15** [comment]
> Hey Muyuan, what's left to do in this PR?

Hey Craig,

Sorry these debug lines are from an accidental commit. I will do another commit for this. 

The optimization for single qubit operations is still a little funky, but other than that I don't think there are much to do for this PR.

**Comment by mlihc15** [comment]
@Strilanc could you help me check what's wrong with pylint in Travis CI? There seems to be an issue with calling pylint.

**Comment by vtomole** [comment]
@mlihc15 Last build succeeded. Was it failing because of the pylint issue you mention? If so, what was the issue?

**Comment by mlihc15** [comment]
> @mlihc15 Last build succeeded. Was it failing because of the pylint issue you mention? If so, what was the issue?

Just quoting Craig: "Pylint is failing because of a bug in the most recent version of its parser astroid. We worked around the issue by pinning the version. Sync to master to get the changes."

**Comment by Strilanc** [comment]
@mlihc15 Just checking: this is ready to review/merge in your opinion?

**Comment by mlihc15** [comment]
> @mlihc15 Just checking: this is ready to review/merge in your opinion?

Yes it is ready for review/merge.

**Comment by vtomole** [comment]
@mlihc15  Ping

**Comment by vtomole** [comment]
@mlihc15 Ping +2

**Comment by c-poole** [comment]
@vtomole It looks like the changes this PR would make were already made in the ion_device merge. The two qubit matrix decomposition into ion trap gates was inserted then (note the several merge conflicts with other cirq/ion files), so this PR can likely be closed as obsolete.

**Pull request closed without merging**